src/sensai/vector_model.py

```diff
@@ -360,41 +405,102 @@ def _fitClassifier(self, X: pd.DataFrame, y: pd.DataFrame):
-        result = [self._labels[i] for i in maxIndices]
-        return pd.DataFrame(result, columns=self.getModelOutputVariableNames())
+        modelOutputVariableName = self.getModelOutputVariableNames()[0]
+        y = self._convertClassProbabilitiesToPredictions(df, predictedVariableName=modelOutputVariableName)
+        y = self._applyPostProcessing(y)
+        return y
 
     def predictClassProbabilities(self, x: pd.DataFrame) -> pd.DataFrame:
         """
         :param x: the input data
         :return: a data frame where the list of columns is the list of class labels and the values are probabilities
         """
-        x = self._computeInputs(x)
+        x = self.computeProcessedInputs(x)
         result = self._predictClassProbabilities(x)
+        self._checkPrediction(result)
+        return result
 
-        # check for correct columns
-        if list(result.columns) != self._labels:
-            raise Exception(f"{self} _predictClassProbabilities returned DataFrame with incorrect columns: expected {self._labels}, got {result.columns}")
+    @abstractmethod
+    def _predictClassProbabilities(self, X: pd.DataFrame) -> pd.DataFrame:
+        pass
 
-        # check for normalisation
-        maxRowsToCheck = 5
-        dfToCheck = result.iloc[:maxRowsToCheck]
-        for i, (_, valueSeries) in enumerate(dfToCheck.iterrows(), start=1):
-            s = valueSeries.sum()
-            if abs(s-1.0) > 0.01:
-                log.warning(f"Probabilities data frame may not be correctly normalised: checked row {i}/{maxRowsToCheck} contains {list(valueSeries)}")
+    def _predict(self, x: pd.DataFrame) -> pd.DataFrame:
+        return self.convertClassProbabilitiesToPredictions(self._predictClassProbabilities(x))
+
+
+class RuleBasedModel(FittableModel, ABC):
+    """
+    Base class for models where the essential prediction logic is based on rules coded by humans
+    and thus does not require fitting. However, the input generation process may use mechanisms
+    such as feature generation or data transformation, which may require fitting.
+    """
+    def __init__(self):
+        super().__init__()
+
+    def fit(self, X: pd.DataFrame, Y: pd.DataFrame = None):
+        self.fitPreprocessors(X, Y=Y)
 
+    def isFitted(self):
+        return self._prePostProcessorsAreFitted()
+
+    def predict(self, x: pd.DataFrame) -> pd.DataFrame:
+        """
+        Performs a prediction for the given input data frame
+
+        :param x: the input data
+        :return: a DataFrame with the same index as the input
+        """
+        x = self.computeProcessedInputs(x)
+        x = self._predict(x)
+        return x
+
+    @abstractmethod
+    def _predict(self, X: pd.DataFrame) -> pd.DataFrame:
+        pass
+
+
+class RuleBasedRegressionModel(RuleBasedModel, ABC):
+    def isRegressionModel(self) -> bool:
+        return True
+
+
+class RuleBasedClassificationModel(ClassificationMixin, RuleBasedModel, ABC):
```

Comment:
There is some code duplication between here and VectorClassificationModel. I could reuse some stuff with the mixin but not all. I think it is bearable and didn't find a better way of organizing the classes. If you have a suggestion for how to improve it, please shoot! :)

Reply:
Outdated